### ISSUE TITLE
Https performance improvements

### ISSUE THREAD

**Issue by DavidPDev** [civetweb/civetweb#887]
Hi all!
We have created a C/C++ server with CivetWeb for our game turn-based server. It must use HTTPS connections, because in Android and iOS it's mandatory (or at least recommended), instead of plain HTTP (which I have also tested).
We have set up everything about certificates etc, and its working perfectly for many months...
Clients (Android, iOS, desktop) just send https requests (GET, POST) polling every 500ms, and all the server logic is done in a Windows Server 2019 dedicated machine.
Now the issue: Some connections (specially far away from server country) have a very high lag (>1000ms) in every https request. I'm almost sure the SSL handshake is being done in every call, and a game may have 100-500 requests.
Any idea on how to impove the lag so just the 1st call does the handshake and the next requests just go "pass through"?
I've done some investigations on the Open SSL and for example you can use the **SSL_set_session** so the session can be reused, but I'm afraid isn't available in CivetWeb.
https://www.openssl.org/docs/man1.0.2/man3/SSL_set_session.html
Also the **SSL_CTX_set_session_cache_mode** may work
https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_session_cache_mode.html

Any ideas? Thanks in advance.

EDIT: We have done a minor improvements:
Removing the 
`SSL_CTX_set_options(dom_ctx->ssl_ctx, SSL_OP_NO_COMPRESSION);`

**Comment by bel2125** [comment]
I'm not sure if SSL_set_session will help here. From my experience, there are a few different scenarios how CivetWeb is used, and optimizations depend on the scenario. Is your scenario close to this:

A client may connect at some time. Once it is connected, they will likely have an interactive session with lots of interactions (~5 per second), and they should be "soft real time" (< 200ms response time in 99.9% of all cases). The numbers are just rough indications / orders of magnitude. Once a client has finished the interactive session, it is either gone completely, or maybe down to ~2 request a minute.
Then optimization depends on the rough number of clients simultaneously in "interactive sessions".

If this number is < 10000, websockets may be a good solution.
If this number is > 100000 ... it must be solved in a different way.

But maybe let's first find out if I got the usage scenario correctly.

**Comment by bel2125** [comment]
Did you already try if there is any effect, if you set `SSL_CTX_set_session_cache_mode(dom_ctx->ssl_ctx, SSL_SESS_CACHE_BOTH);`? 

I found only one code example for `SSL_CTX_set_session_cache_mode`, and this does not require `SSL_set_session`: https://www.w3.org/Library/src/SSL/HTSSL.c ... but, have a look at the comment: 
```C
	/* Not sure what this does */
        SSL_CTX_set_session_cache_mode(app_ctx, SSL_SESS_CACHE_CLIENT);
```

**Comment by DavidPDev** [comment]
Thank you very much for the quick reply!

I will do some more tests about the SSL_CTX_set_session_cache_mode(dom_ctx->ssl_ctx, SSL_SESS_CACHE_BOTH); (I test it already this morning without any effect).

I have just added  the following optimization that seems to improve a lot the "lag":

`"keep_alive_timeout_ms",		"10000",		// 10s. of connection re-use`

The default value for the keep_alive timeout was 500ms just the same polling value I was using so connections where never "keeped-alive" :(

Also increased the "num_threads" to "1000".

We have just launched the server to LIVE and most of the users are geting pings < 100ms. Great improvement!

@bel2125 : Again thanks for your great work on Civetweb!

**Comment by bel2125** [comment]
> I will do some more tests about the SSL_CTX_set_session_cache_mode(dom_ctx->ssl_ctx, SSL_SESS_CACHE_BOTH); (I test it already this morning without any effect).

Activating session caching in a server seems to be a security issue, if you do not restart the server regularly:
https://www.postgresql.org/message-id/22227.1501632275%40sss.pgh.pa.us

The sequence to activate it seems to be `SSL_CTX_set_session_id_context`, `SSL_CTX_set_timeout` and finally `SSL_CTX_set_session_cache_mode`. The `SSL_CTX_set_session_id_context` needs a 32 byte session identifier that should be recreated after a while.


> "keep_alive_timeout_ms", "10000", // 10s. of connection re-use

Might be the 500 ms default is too small for high ping rates of intercontinental connections.


> Also increased the "num_threads" to "1000".

This also limits the maximum number of concurrent connections, including those in a "waiting" / "closing" / .. state.
I did tests with 30000, others tested with more. A dedicated server can certainly handle more than 10000.

If you are using websockets, you could also keep this number of client connections open for a bidirectional communication without any new handshake. I don't know if you are aware how websockets work. After an initial HTTP/HTTPS handshake, the connection remains open and you can exchange messages in both directions. Websocket connections can remain open for an unlimited time (although you should send a message every 30s / according to the configuration, or you can use PING/PONG messages), there is no additional handshake one the connection is established. To me websockets seem to be an ideal solution to real-time interactive pages (games).
But this requires a change in the client program - it's not just server tuning.

**Comment by DavidPDev** [comment]
Thank you again for all those insights!
For now, we will continue testing the SSL_CTX_session_xxxx related functions.
Also increasing/tweaking the keep_alive timeout and number of threads.
Definitely websockets should be the optimal solution for any kind of fluent client<->server communication, we'll keep the work on that direction, but we need to do many changes on server + client.

**Comment by bel2125** [comment]
> For now, we will continue testing the SSL_CTX_session_xxxx related functions.

I would appreciate if you share some insights from your tests.
Probably SSL_CTX_session_xxxx could be a beneficial feature to add to the standard.

**Comment by DavidPDev** [comment]
Lastest tests:
In the `init_ssl_ctx_impl ()` body function added the following lines after `SSL_CTX_set_session_id_context ...`

```
SSL_CTX_set_session_cache_mode(dom_ctx->ssl_ctx, SSL_SESS_CACHE_BOTH);
SSL_CTX_sess_set_cache_size(dom_ctx->ssl_ctx, 10000);
SSL_CTX_set_timeout(dom_ctx->ssl_ctx, 60 * 60L);
```
Also set threads to 10000

I have to do some more tests and intercontinental users measurements but the users lags have been reduced at least a **70%** (in mean) or even more. :)
Now most pings are 60-150 ms, before where 250-600 ms.

Tomorrow I will have more data.

**Comment by bel2125** [comment]
I added an implementation for SSL session caching - it seems to work, but I did not do any performance tests.
There is a new configuration value "ssl_cache_timeout" - if this value is set to some time, session caching is activated with the configured timeout. Default value is "-1" (no caching). 
SSL_CTX_sess_set_cache_size is currently not set in the code - I just used the default.
It is already available in the master branch will be available in the next release in a couple of days.
